### PR TITLE
#463 filter review home by stage and reviewer

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #463 

### Changes

* make sure stage id is available in full structure stage list
* add `< ReviewerAndStageSelection>` to show two drop downs, one for stage and one for reviewer
* add filters to `<ReviewHome>` to be set by `<ReviewerAndStageSelection>`

Was trying to not use useEffect, but it wold have required making `ReviewerAndStageSelection` a wrapper, since fullStructure is not available at the start

### Testing

Try with 34567/review

### CODE REVIEW

File: src/containers/Review/ReviewHome.tsx [hunk 1]
```diff
@@ -55,10 +66,96 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
...
+  const changeFilters = (filterType: keyof Filters) => (_: any, { value }: any) => {
+    if (filters) setFilters({ ...filters, [filterType]: value })
+  }
+
+  if (!filters) return null
+
+  return (
+    <Grid columns="equal">
+      <Grid.Column floated="left">
+        Show tasks for{' '}
```
Comment (nmadruga) [inline src/containers/Review/ReviewHome.tsx:108]: Plase localise : )

File: src/containers/Review/ReviewHome.tsx [hunk 2]
```diff
@@ -55,10 +66,96 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
...
+    if (id === currentUserId || !id || !firstName || !lastName) return
+    reviewerOptions.push({
+      value: id,
+      key: id,
+      text: `${firstName} ${lastName}`,
+    })
+  })
+
+  return reviewerOptions
+}
```
Comment (nmadruga) [inline src/containers/Review/ReviewHome.tsx:159]: I think those 2 should be moved to some helper function - for better readability of this. The name of the function already tells the story

File: src/containers/Review/ReviewHome.tsx [hunk 3]
```diff
@@ -55,10 +66,96 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
...
+      value: id,
+      text: title,
+    }))
+
+const getReviewerOptions = (assignments: AssignmentDetailsNEW[], currentUserId: number) => {
+  const reviewerOptions: { value: number; key: number; text: string }[] = [
+    {
+      value: ALL_REVIEWERS,
+      key: ALL_REVIEWERS,
+      text: 'All',
```
Comment (nmadruga) [inline src/containers/Review/ReviewHome.tsx:141]: To localise 2

File: src/containers/Review/ReviewHome.tsx [hunk 4]
```diff
@@ -55,10 +66,96 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
...
+    },
+    {
+      value: currentUserId,
+      key: currentUserId,
+      text: 'Yourself',
```
Comment (nmadruga) [inline src/containers/Review/ReviewHome.tsx:146]: To localise 3

File: src/containers/Review/ReviewHome.tsx [hunk 5]
```diff
@@ -55,10 +66,96 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
...
+      <Grid.Column floated="left">
+        Show tasks for{' '}
+        <Dropdown
+          options={getReviewerOptions(assignments, currentUser?.userId as number)}
+          value={filters?.selectedReviewer}
+          onChange={changeFilters('selectedReviewer')}
+        />
+      </Grid.Column>
+      <Grid.Column floated="right" textAlign="right">
+        STAGE{' '}
```
Comment (nmadruga) [inline src/containers/Review/ReviewHome.tsx:116]: To localise 4

**Review by nmadruga — COMMENTED** [review]
Just a few comments. Testing now

**Comment by nmadruga** [comment]
Hmm testing the UI it has some non-expected behaviours. Selecting something on either one of the dropdown not always show the correct result... I wonder if this one is going to have a Part 2. Also should we use URL filter to display, what is the benefit in this case - I think the bugs might be related to page reload and selection of item in dropdown. Not sure, just made a quick test.

File: src/containers/Review/ReviewHome.tsx (hunk 1 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: sorry, missed it: 92a4c9992b32a2d8370573f6d842417c595e4c94

File: src/containers/Review/ReviewHome.tsx (hunk 2 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: sorry, missed it: 92a4c99

File: src/containers/Review/ReviewHome.tsx (hunk 3 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: sorry, missed it: 92a4c99

File: src/containers/Review/ReviewHome.tsx (hunk 4 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: sorry, missed it: 92a4c99

File: src/containers/Review/ReviewHome.tsx (hunk 5 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: 75ee2014fa6a0379e4fc44773462dbf921da3088

File: src/containers/Review/ReviewHome.tsx (hunk 2 above)
Comment (andreievg) [reply src/containers/Review/ReviewHome.tsx]: Hmm, I was also thinking it's getting a bit much for ReviewHome page. What do you think, if i move the whole component `ReviewerAndStageSelection` along with it's helper methods, would that be ok ?

**Comment by andreievg** [comment]
> Hmm testing the UI it has some non-expected behaviours. Selecting something on either one of the dropdown not always show the correct result... I wonder if this one is going to have a Part 2. Also should we use URL filter to display, what is the benefit in this case - I think the bugs might be related to page reload and selection of item in dropdown. Not sure, just made a quick test.

I am glad you pointed this out. Even thought my testing was not showing any bugs, but I've tried a bit harder:

After clearing cache, when selecting all there is a spasm (then the spasm goes away).

I was looking at network request (btw added this to see what operations are firing): 0ce66d56e9fb601346d1138fe2c1ecaf69776682

And every time after cache is reset, reviewGetInfo is re-fired (in the middle of getting reviewResponses), this was very similar to what we were seeing with totalCount on questionAssignments (except it was just causing spasms, not UI freezes).

I ended up figuring it out, for some reason, filtering of reviews in reviewResponse query, causes reviewGetInfo to re-fire (it must notice that different review.nodes are returned for reviewAssignment, so decided getReviewInfo needs to be re-fried). I think it's all fixed now (by filtering at reviewResponse not review end, as it should be anyway, to allow other reviewers looking at previously submitted reviewResponse):

15f85f84945fda9b0a12a21dda01bb4ca9db611a

File: src/containers/Review/ReviewHome.tsx (hunk 2 above)
Comment (nmadruga) [reply src/containers/Review/ReviewHome.tsx]: Yes, that will be nice

**Comment by nmadruga** [comment]
Awesome! The feature seems to work very nicely now.

**Review by nmadruga — APPROVED** [review]
👍 Thanks for all changes and new feature.

Have just one minor comment to make - but something that can be changed afterwards (part of #356 maybe). It's not important for now. I noticed the page doesn't show 404 when Reviewer lvl 1 isn't assigned to something. I think it's because they can still see Assigned to All options. But I wonder if this shouldn't be visible (or if permissions would prevent this).

The page showing when loaded (no reviews assigned to Reviewer2):
Screen Shot 2021-03-21 at 5 54 32 PM


And when filtering assigned to ALL:
Screen Shot 2021-03-21 at 5 54 21 PM

**Comment by andreievg** [comment]
Thanks fore review @nmadruga

**Comment by andreievg** [comment]
> 👍 Thanks for all changes and new feature.

> Have just one minor comment to make - but something that can be changed afterwards (part of #356 maybe). It's not important for now. I noticed the page doesn't show 404 when Reviewer lvl 1 isn't assigned to something. I think it's because they can still see Assigned to All options. But I wonder if this shouldn't be visible (or if permissions would prevent this).

> The page showing when loaded (no reviews assigned to Reviewer2):

Hmm, at the very least it would be good to parameterise stage and user select (navigate to 'all', if user has no tasks for the application, and perhaps change stage to latest stage to latest stage of assignments they can see)

**Comment by andreievg** [comment]
https://github.com/openmsupply/application-manager-web-app/issues/485

**Pull request merged**